genienlp/tasks/generic_task.py

```diff
@@ -253,8 +253,65 @@ def get_splits(self, root, **kwargs):
         return OODDataset.splits(root=root, **kwargs)
 
 
+class E2EDialogueTask(BaseTask):
+    def __init__(self, name, args):
+        super().__init__(name, args)
+        self._metrics = ['e2e_dialogue_score']
+
+    def utterance_field(self):
+        return 'context'
+
+    def _make_example(self, turn, **kwargs):
+        dial_id, turn_id, input_text, output_text, train_target = (
+            turn['dial_id'],
+            turn['turn_id'],
+            turn['input_text'],
+            turn['output_text'],
+            turn['train_target'],
+        )
+
+        if kwargs.get('train_target', False) and train_target != kwargs['train_target']:
+            return None
+
+        example_id = '/'.join([dial_id, str(turn_id), train_target])
+
+        return Example.from_raw(
+            self.name + '/' + str(example_id), input_text, '', output_text, preprocess=self.preprocess_field, lower=False
+        )
+
+    def get_splits(self, root, **kwargs):
+        kwargs['e2e_evaluation'] = self.args.e2e_dialogue_evaluation
+        return E2EDialogueDataset.return_splits(path=root, make_example=self._make_example, **kwargs)
+
+
+@register_task('risawoz')
+class RiSAWOZ(E2EDialogueTask):
+    def __init__(self, name, args):
+        super().__init__(name, args)
+        self.special_tokens = {
```

Comment:
I think we have two different `special_tokens` in GenieNLP:

1. The `--preprocess_special_tokens` argument in `arguments.py` and related code which means we replace long ThingTalk tokens with a smaller prefix as a preprocessing step.
1. The `special_tokens` here, which are supposed to be treated as one token by the tokenizer?

Am I understanding this correctly? If so, why use the same name?

Reply:
Yeah `special_tokens` name is a bit overloaded in the code, but we are essentially using it for any token that should not be touched by the tokenizer. `--preprocess_special_tokens` should really be `--preprocess_thingtalk_tokens`. What it does is breaking thingtalk tokens into smaller tokens that are then treated as special tokens.